yt_idefix/data_structures.py

```diff
@@ -427,15 +427,173 @@ def _get_time(self):
 
     def _set_code_unit_attributes(self):
         """Conversion between physical units and code units."""
-        length_unit = self.parameters.get("unit_length", 1.0)
-        velocity_unit = self.parameters.get("unit_velocity", 1.0)
-        density_unit = self.parameters.get("unit_density", 1.0)
 
-        self.length_unit = self.quan(length_unit, "cm")
-        self.velocity_unit = self.quan(velocity_unit, "cm/s")
-        density_unit = self.quan(density_unit, "g/cm**3")
+        # User-passed units are preferred to PLUTO's base units for deriving all units
+        pluto_units = {
+            "length_unit": "cm",
+            "velocity_unit": "cm/s",
+            "density_unit": "g/cm**3",
+        }
+        spu = set(pluto_units)
+        uo = self.units_override.copy()
+        if len(uo) < 3:
+            if len(uo) > 0:
+                ytLogger.info(
+                    "Less than 3 units were specified in units_override (got %s). "
+                    "Need to rely on PLUTO's internal units to derive other units",
+                    len(uo),
+                )
+            # Remove those already in units_override
+            spu.difference_update(uo)
+            for unit in spu:
+                uo[unit] = self.quan(self.parameters.get(unit, 1.0), pluto_units[unit])
+                if len(self.units_override) > 0:
+                    try:
+                        self._validate_units_override_keys(uo)
+                        ytLogger.info("Relying on %s: %s.", unit, uo[unit])
+                    except ValueError:
+                        del uo[unit]
+                if len(uo) == 3:
+                    break
+
+        # Then we calculate base units via the units in units_override
+        # The condition statements are essential to prevent getting stuck in infinite recursion
+        # However they are fragile, you must be cautious to change them
+        def _time():
+            if "time_unit" not in uo:
+                uo["time_unit"] = uo["length_unit"] / uo["velocity_unit"]
```

Comment:
I guess we could lower the indentation level by using guard conditions instead, such as
```suggestion
            if "time_unit" in uo:
                return
            uo["time_unit"] = uo["length_unit"] / uo["velocity_unit"]
```